File: jscomp/others/bs_SortedMapDict.ml
(* Copyright (C) 2017 Authors of BuckleScript
 * 
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * In addition to the permissions granted to you by the LGPL, you may combine
 * or link a "work that uses the Library" with a publicly distributed version
 * of this file to produce a combined library or application, then distribute
 * that combined work under the terms of your choosing, with no requirement
 * to comply with the obligations normally placed on you by section 4 of the
 * LGPL version 3 (or the corresponding section of a later version of the LGPL
 * should you choose to use a later version).
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 * 
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA. *)

module N = Bs_internalAVLtree
module A = Bs_Array

type ('key,  'a, 'id) t = ('key, 'a) N.t0

type ('key, 'id) cmp = ('key, 'id)  Bs_Cmp.cmp

let empty = N.empty0
let ofArray = N.ofArray0
let isEmpty = N.isEmpty0
let cmp = N.cmp0
let eq = N.eq0   
let has = N.mem0
let forEach = N.iter0      
let reduce = N.fold0
let every = N.every0
let some = N.some0    
let size = N.length0
let toList = N.toList0
let toArray = N.toArray0
let keysToArray = N.keysToArray0
let valuesToArray = N.valuesToArray0

let minimum = N.minKVOpt0
let maximum = N.maxKVOpt0
let minKey = N.minKeyOpt0
let maxKey = N.maxKeyOpt0
let minKeyNull = N.minKeyNull0
let maxKeyNull = N.maxKeyNull0
let minNull = N.minKVNull0
let maxNull = N.maxKVNull0
let get = N.findOpt0
let getNull = N.findNull0
let getWithDefault = N.findWithDefault0
let getExn = N.findExn0

let mapWithKey = N.mapi0
let map  = N.map0

let keepBy = N.filterShared0
let partition = N.partitionShared0
let checkInvariantInternal = N.checkInvariantInternal
let rec set  (t : _ t) newK newD  ~cmp =
  match N.toOpt t with 
  | None -> N.singleton0 newK newD 
  | Some n  ->
    let k= N.key n in 
    let c = (Bs_Cmp.getCmp cmp) newK k [@bs] in
    if c = 0 then
      N.return (N.updateValue n newD) 
    else 
      let l,r,v = N.left n, N.right n, N.value n in 
      if c < 0 then (* Worth optimize for reference equality? *)
        N.bal (set ~cmp l newK newD ) k v  r
      else
        N.bal l k v (set ~cmp r newK newD )

let rec update  (t : _ t) newK f  ~cmp :  _ t =
  match N.toOpt t with 
  | None ->
    begin match f None [@bs] with 
      | None -> t 
      | Some newD -> N.singleton0 newK newD 
    end 
  | Some n  ->
    let k= N.key n in 
    let c = (Bs_Cmp.getCmp cmp) newK k [@bs] in
    if c = 0 then
      match f (Some (N.value n)) [@bs] with 
      | None ->
        let l, r = N.left n , N.right n in  
        begin match N.toOpt l, N.toOpt r with
        | None, _ -> r
        | _, None -> l
        | _, Some rn ->
          let kr, vr = ref (N.key rn), ref (N.value rn) in
          let r = N.removeMinAuxWithRef rn kr vr in
          N.bal l !kr !vr r 
        end
      | Some newD -> N.return (N.updateValue n newD)
    else 
      let l,r,v = N.left n, N.right n, N.value n in 
      if c < 0 then
        let ll = (update ~cmp l newK f ) in
        if l == ll then
          t
        else 
          N.bal ll k v  r            
      else
        let rr = (update ~cmp r newK f) in
        if r == rr then t 
        else N.bal l k v rr

(*  unboxing API was not exported
    since the correct API is really awkard
    [bool -> 'k Js.null -> ('a Js.null * bool)]
    even for specialized [k] the first [bool] can 
    be erased, maybe the perf boost does not justify the inclusion of such API

    [updateWithNull m x f]
    the callback to [f exist v] 
    when [v] is non-null,
    [exist] is guaranteed to be true
    [v] is guranteed to be [null],
    when [exist] is [true], [v] could be [null], 
    since ['a] is polymorphic
*)


let rec removeAux0  n x ~cmp = 
  let l,v,r = N.(left n, key n, right n ) in 
  let c = (Bs_Cmp.getCmp cmp) x v [@bs] in
  if c = 0 then
    match N.toOpt l, N.toOpt r with 
    | None, _ -> r 
    | _, None -> l 
    | _, Some rn -> 
      let kr, vr = ref (N.key rn), ref (N.value rn) in 
      let r = N.removeMinAuxWithRef rn kr vr in 
      N.bal l !kr !vr r
  else if c < 0 then
    match N.toOpt l with 
    | None -> N.return n (* Nothing to remove *)
    | Some left ->
      let ll = removeAux0 left x ~cmp in 
      if ll == l then (N.return n)
      else N.bal ll v (N.value n) r
  else
    match N.toOpt r with 
    | None -> N.return n (* Nothing to remove *)
    | Some right -> 
      let rr = removeAux0 ~cmp right x in
      if rr == r then N.return n
      else N.bal l v (N.value n)  rr


let remove n x ~cmp = 
  match N.toOpt n with        
  | None -> N.empty0
  | Some n -> removeAux0 n x ~cmp 
  
let mergeMany   h arr ~cmp =   
  let len = A.length arr in 
  let v = ref h in  
  for i = 0 to len - 1 do 
    let key,value = A.getUnsafe arr i in 
    v := set !v  ~cmp key value
  done ;
  !v 

let rec splitAuxPivot n x pres  ~cmp =  
  let l,v,d,r = N.(left n , key n, value n, right n) in  
  let c = (Bs_Cmp.getCmp cmp) x v [@bs] in 
  if c = 0 then begin 
    pres := Some d; 
    (l,  r)
  end
  else     
  if c < 0 then
    match N.toOpt l with 
    | None -> 
      N.(empty, return n)
    | Some l -> 
      let (ll,rl) = splitAuxPivot ~cmp l x pres in
      (ll,  N.join rl v d r)
  else
    match N.toOpt r with 
    | None ->
      N.(return n, empty)
    | Some r -> 
      let (lr,  rr) = splitAuxPivot ~cmp r x pres in
      (N.join l v d lr,  rr)


let split  n x ~cmp = 
  match N.toOpt n with 
  | None ->     
    N.(empty, empty), None
  | Some n  ->
    let pres = ref None in
    let v = splitAuxPivot ~cmp n x pres in 
    v, !pres

let rec merge s1 s2 f ~cmp =
  match N.(toOpt s1, toOpt s2) with
    (None, None) -> N.empty
  | Some _, None -> 
    N.filterMap0 s1 (fun[@bs] k v -> 
        f k (Some v) None [@bs]
      )
  | None, Some _ -> 
    N.filterMap0 s2 (fun[@bs] k v -> 
        f k None (Some v) [@bs]
      )
  | Some s1n , Some s2n -> 
    if N.h s1n  >= N.h s2n  then
      let l1, v1, d1, r1 = N.(left s1n, key s1n, value s1n, right s1n) in 
      let d2 = ref None in 
      let (l2, r2) = splitAuxPivot ~cmp s2n v1 d2 in
      let d2 = !d2 in 
      let newLeft = merge ~cmp l1 l2 f in 
      let newD = f v1 (Some d1) d2 [@bs] in 
      let newRight = merge ~cmp r1 r2 f in 
      N.concatOrJoin newLeft v1 newD  newRight
    else
      let l2,v2,d2,r2 = N.(left s2n, key s2n, value s2n, right s2n) in 
      let d1 = ref None in 
      let (l1,  r1) = splitAuxPivot ~cmp s1n v2 d1 in
      let d1 = !d1 in 
      let newLeft = merge ~cmp l1 l2 f in 
      let newD = (f v2 d1 (Some d2) [@bs]) in 
      let newRight = (merge ~cmp r1 r2 f) in 
      N.concatOrJoin newLeft v2 newD newRight

let rec removeArrayAux t xs i len ~cmp =
  if i < len then
    let ele = A.getUnsafe xs i in
    let u =  removeAux0 t ele ~cmp in
    match N.toOpt u with
    | None -> u
    | Some t -> removeArrayAux t xs (i + 1) len ~cmp 
  else
    N.return t
      
let removeMany t keys ~cmp =
  let len = A.length keys in
  match N.toOpt t with
  | None -> N.empty0
  | Some t ->  removeArrayAux t keys 0 len ~cmp 

File: lib/js/bs_SortedMapDict.js
'use strict';

var Bs_internalAVLtree = require("./bs_internalAVLtree.js");

function set(t, newK, newD, cmp) {
  if (t !== null) {
    var k = t.key;
    var c = cmp(newK, k);
    if (c) {
      var l = t.left;
      var r = t.right;
      var v = t.value;
      if (c < 0) {
        return Bs_internalAVLtree.bal(set(l, newK, newD, cmp), k, v, r);
      } else {
        return Bs_internalAVLtree.bal(l, k, v, set(r, newK, newD, cmp));
      }
    } else {
      return Bs_internalAVLtree.updateValue(t, newD);
    }
  } else {
    return Bs_internalAVLtree.singleton0(newK, newD);
  }
}

function update(t, newK, f, cmp) {
  if (t !== null) {
    var k = t.key;
    var c = cmp(newK, k);
    if (c) {
      var l = t.left;
      var r = t.right;
      var v = t.value;
      if (c < 0) {
        var ll = update(l, newK, f, cmp);
        if (l === ll) {
          return t;
        } else {
          return Bs_internalAVLtree.bal(ll, k, v, r);
        }
      } else {
        var rr = update(r, newK, f, cmp);
        if (r === rr) {
          return t;
        } else {
          return Bs_internalAVLtree.bal(l, k, v, rr);
        }
      }
    } else {
      var match = f(/* Some */[t.value]);
      if (match) {
        return Bs_internalAVLtree.updateValue(t, match[0]);
      } else {
        var l$1 = t.left;
        var r$1 = t.right;
        if (l$1 !== null) {
          if (r$1 !== null) {
            var kr = [r$1.key];
            var vr = [r$1.value];
            var r$2 = Bs_internalAVLtree.removeMinAuxWithRef(r$1, kr, vr);
            return Bs_internalAVLtree.bal(l$1, kr[0], vr[0], r$2);
          } else {
            return l$1;
          }
        } else {
          return r$1;
        }
      }
    }
  } else {
    var match$1 = f(/* None */0);
    if (match$1) {
      return Bs_internalAVLtree.singleton0(newK, match$1[0]);
    } else {
      return t;
    }
  }
}

function removeAux0(n, x, cmp) {
  var l = n.left;
  var v = n.key;
  var r = n.right;
  var c = cmp(x, v);
  if (c) {
    if (c < 0) {
      if (l !== null) {
        var ll = removeAux0(l, x, cmp);
        if (ll === l) {
          return n;
        } else {
          return Bs_internalAVLtree.bal(ll, v, n.value, r);
        }
      } else {
        return n;
      }
    } else if (r !== null) {
      var rr = removeAux0(r, x, cmp);
      if (rr === r) {
        return n;
      } else {
        return Bs_internalAVLtree.bal(l, v, n.value, rr);
      }
    } else {
      return n;
    }
  } else if (l !== null) {
    if (r !== null) {
      var kr = [r.key];
      var vr = [r.value];
      var r$1 = Bs_internalAVLtree.removeMinAuxWithRef(r, kr, vr);
      return Bs_internalAVLtree.bal(l, kr[0], vr[0], r$1);
    } else {
      return l;
    }
  } else {
    return r;
  }
}

function remove(n, x, cmp) {
  if (n !== null) {
    return removeAux0(n, x, cmp);
  } else {
    return Bs_internalAVLtree.empty0;
  }
}

function mergeMany(h, arr, cmp) {
  var len = arr.length;
  var v = h;
  for(var i = 0 ,i_finish = len - 1 | 0; i <= i_finish; ++i){
    var match = arr[i];
    v = set(v, match[0], match[1], cmp);
  }
  return v;
}

function splitAuxPivot(n, x, pres, cmp) {
  var l = n.left;
  var v = n.key;
  var d = n.value;
  var r = n.right;
  var c = cmp(x, v);
  if (c) {
    if (c < 0) {
      if (l !== null) {
        var match = splitAuxPivot(l, x, pres, cmp);
        return /* tuple */[
                match[0],
                Bs_internalAVLtree.join(match[1], v, d, r)
              ];
      } else {
        return /* tuple */[
                null,
                n
              ];
      }
    } else if (r !== null) {
      var match$1 = splitAuxPivot(r, x, pres, cmp);
      return /* tuple */[
              Bs_internalAVLtree.join(l, v, d, match$1[0]),
              match$1[1]
            ];
    } else {
      return /* tuple */[
              n,
              null
            ];
    }
  } else {
    pres[0] = /* Some */[d];
    return /* tuple */[
            l,
            r
          ];
  }
}

function split(n, x, cmp) {
  if (n !== null) {
    var pres = [/* None */0];
    var v = splitAuxPivot(n, x, pres, cmp);
    return /* tuple */[
            v,
            pres[0]
          ];
  } else {
    return /* tuple */[
            /* tuple */[
              null,
              null
            ],
            /* None */0
          ];
  }
}

function merge(s1, s2, f, cmp) {
  if (s1 !== null) {
    if (s2 !== null) {
      if (s1.h >= s2.h) {
        var l1 = s1.left;
        var v1 = s1.key;
        var d1 = s1.value;
        var r1 = s1.right;
        var d2 = [/* None */0];
        var match = splitAuxPivot(s2, v1, d2, cmp);
        var d2$1 = d2[0];
        var newLeft = merge(l1, match[0], f, cmp);
        var newD = f(v1, /* Some */[d1], d2$1);
        var newRight = merge(r1, match[1], f, cmp);
        return Bs_internalAVLtree.concatOrJoin(newLeft, v1, newD, newRight);
      } else {
        var l2 = s2.left;
        var v2 = s2.key;
        var d2$2 = s2.value;
        var r2 = s2.right;
        var d1$1 = [/* None */0];
        var match$1 = splitAuxPivot(s1, v2, d1$1, cmp);
        var d1$2 = d1$1[0];
        var newLeft$1 = merge(match$1[0], l2, f, cmp);
        var newD$1 = f(v2, d1$2, /* Some */[d2$2]);
        var newRight$1 = merge(match$1[1], r2, f, cmp);
        return Bs_internalAVLtree.concatOrJoin(newLeft$1, v2, newD$1, newRight$1);
      }
    } else {
      return Bs_internalAVLtree.filterMap0(s1, (function (k, v) {
                    return f(k, /* Some */[v], /* None */0);
                  }));
    }
  } else if (s2 !== null) {
    return Bs_internalAVLtree.filterMap0(s2, (function (k, v) {
                  return f(k, /* None */0, /* Some */[v]);
                }));
  } else {
    return null;
  }
}

function removeMany(t, keys, cmp) {
  var len = keys.length;
  if (t !== null) {
    var _t = t;
    var xs = keys;
    var _i = 0;
    var len$1 = len;
    var cmp$1 = cmp;
    while(true) {
      var i = _i;
      var t$1 = _t;
      if (i < len$1) {
        var ele = xs[i];
        var u = removeAux0(t$1, ele, cmp$1);
        if (u !== null) {
          _i = i + 1 | 0;
          _t = u;
          continue ;
          
        } else {
          return u;
        }
      } else {
        return t$1;
      }
    };
  } else {
    return Bs_internalAVLtree.empty0;
  }
}

var empty = Bs_internalAVLtree.empty0;

var isEmpty = Bs_internalAVLtree.isEmpty0;

var has = Bs_internalAVLtree.mem0;

var cmp = Bs_internalAVLtree.cmp0;

var eq = Bs_internalAVLtree.eq0;

var forEach = Bs_internalAVLtree.iter0;

var reduce = Bs_internalAVLtree.fold0;

var every = Bs_internalAVLtree.every0;

var some = Bs_internalAVLtree.some0;

var size = Bs_internalAVLtree.length0;

var toList = Bs_internalAVLtree.toList0;

var toArray = Bs_internalAVLtree.toArray0;

var ofArray = Bs_internalAVLtree.ofArray0;

var keysToArray = Bs_internalAVLtree.keysToArray0;

var valuesToArray = Bs_internalAVLtree.valuesToArray0;

var minKey = Bs_internalAVLtree.minKey0;

var minKeyUndefined = Bs_internalAVLtree.minKeyUndefined0;

var maxKey = Bs_internalAVLtree.maxKey0;

var maxKeyUndefined = Bs_internalAVLtree.maxKeyUndefined0;

var minimum = Bs_internalAVLtree.minimum0;

var minUndefined = Bs_internalAVLtree.minUndefined0;

var maximum = Bs_internalAVLtree.maximum0;

var maxUndefined = Bs_internalAVLtree.maxUndefined0;

var get = Bs_internalAVLtree.get0;

var getUndefined = Bs_internalAVLtree.getUndefined0;

var getWithDefault = Bs_internalAVLtree.getWithDefault0;

var getExn = Bs_internalAVLtree.getExn0;

var checkInvariantInternal = Bs_internalAVLtree.checkInvariantInternal;

var keepBy = Bs_internalAVLtree.filterShared0;

var partition = Bs_internalAVLtree.partitionShared0;

var map = Bs_internalAVLtree.map0;

var mapWithKey = Bs_internalAVLtree.mapi0;

exports.empty = empty;
exports.isEmpty = isEmpty;
exports.has = has;
exports.cmp = cmp;
exports.eq = eq;
exports.forEach = forEach;
exports.reduce = reduce;
exports.every = every;
exports.some = some;
exports.size = size;
exports.toList = toList;
exports.toArray = toArray;
exports.ofArray = ofArray;
exports.keysToArray = keysToArray;
exports.valuesToArray = valuesToArray;
exports.minKey = minKey;
exports.minKeyUndefined = minKeyUndefined;
exports.maxKey = maxKey;
exports.maxKeyUndefined = maxKeyUndefined;
exports.minimum = minimum;
exports.minUndefined = minUndefined;
exports.maximum = maximum;
exports.maxUndefined = maxUndefined;
exports.get = get;
exports.getUndefined = getUndefined;
exports.getWithDefault = getWithDefault;
exports.getExn = getExn;
exports.checkInvariantInternal = checkInvariantInternal;
exports.remove = remove;
exports.removeMany = removeMany;
exports.set = set;
exports.update = update;
exports.merge = merge;
exports.mergeMany = mergeMany;
exports.keepBy = keepBy;
exports.partition = partition;
exports.split = split;
exports.map = map;
exports.mapWithKey = mapWithKey;
/* No side effect */
